### PR TITLE
executor: improve SHOW commands compatibility

### PR DESCRIPTION
**Pull request by morgo**
### What problem does this PR solve? 

Improves MySQL compatibility of SHOW commands, plus an enhancement to derive info from infoschema to reduce duplication.

### What is changed and how it works?

Lots of small compatibility bugs:

- Map `SHOW ENGINES` and `SHOW TABLE STATUS` to infoschema (saves duplication)
- Fix return type of `SHOW TABLE STATUS` column
- Add privilege checks before checking if DB exists as in MySQL.
- Use correct MySQL error codes for errors, privilege checks
- Added partition info to infoschema.

### Check List 

Tests 

 - Unit test
 - Manual test

Code changes

- Minimal

Side effects

 - Breaking backward compatibility

The added privilege checks may break incorrect apps.

Related changes

 - Need to update the `tidb-ansible` repository
 - Need to be included in the release note

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8417)

### CODE REVIEW

**Comment by tiancaiamao** [comment]
LGTM

File: executor/show.go [hunk 1]
```diff
@@ -155,14 +156,16 @@ func (e *ShowExec) fetchAll() error {
...
 func (e *ShowExec) fetchShowEngines() error {
-	e.appendRow([]interface{}{
-		"InnoDB",
-		"DEFAULT",
-		"Supports transactions, row-level locking, and foreign keys",
-		"YES",
-		"YES",
-		"YES",
-	})
+
```
Comment (zz-jason) [inline executor/show.go]: remove this empty line?

File: executor/show.go [hunk 2]
```diff
@@ -241,10 +244,13 @@ func (e *ShowExec) fetchShowProcessList() error {
 }
 
 func (e *ShowExec) fetchShowTables() error {
+	checker := privilege.GetPrivilegeManager(e.ctx)
+	if checker != nil && e.ctx.GetSessionVars().User != nil && !checker.DBIsVisible(fmt.Sprint(e.DBName)) {
```
Comment (zz-jason) [inline executor/show.go]: s/fmt.Sprint(e.DBName)/e.DBName/

File: executor/show.go [hunk 3]
```diff
@@ -269,18 +275,34 @@ func (e *ShowExec) fetchShowTables() error {
 }
 
 func (e *ShowExec) fetchShowTableStatus() error {
+	checker := privilege.GetPrivilegeManager(e.ctx)
+	if checker != nil && e.ctx.GetSessionVars().User != nil && !checker.DBIsVisible(fmt.Sprint(e.DBName)) {
```
Comment (zz-jason) [inline executor/show.go]: ditto

**Pull request merged**